syntax = "proto2";

package mahjong;

enum TileType {
  // 0x00 00 00 00 00 00 00 00
  //      G  F  E  D  C  B  A
  //
  // A: [1, 9] number of a tile. Or 0 means any.
  //
  // B: [1, 5] tile type. Or 0 means any.
  //
  // C: [1, 2] sequential or not. Or 0 means any.
  MASK_TILE_NUMBER = 0xff;
  MASK_TILE_KIND = 0xff00;
  MASK_TILE_SEQUENTIAL = 0xff0000;

  WIND_TILE   = 0x0100;
  SANGEN_TILE = 0x0200;
  WANZU_TILE  = 0x0300;
  SOUZU_TILE  = 0x0400;
  PINZU_TILE  = 0x0500;

  UNSEQUENTIAL_TILE = 0x010000;
  SEQUENTIAL_TILE   = 0x020000;

  JIHAI_TILE = 0x010000;

  WIND_TON = 0x010101;
  WIND_NAN = 0x010102;
  WIND_SHA = 0x010103;
  WIND_PE  = 0x010104;

  SANGEN_HAKU  = 0x010201;
  SANGEN_HATSU = 0x010202;
  SANGEN_CHUN  = 0x010203;

  TILE_1 = 0x020001;
  TILE_2 = 0x020002;
  TILE_3 = 0x020003;
  TILE_4 = 0x020004;
  TILE_5 = 0x020005;
  TILE_6 = 0x020006;
  TILE_7 = 0x020007;
  TILE_8 = 0x020008;
  TILE_9 = 0x020009;

  WANZU_1 = 0x020301;
  WANZU_2 = 0x020302;
  WANZU_3 = 0x020303;
  WANZU_4 = 0x020304;
  WANZU_5 = 0x020305;
  WANZU_6 = 0x020306;
  WANZU_7 = 0x020307;
  WANZU_8 = 0x020308;
  WANZU_9 = 0x020309;

  SOUZU_1 = 0x020401;
  SOUZU_2 = 0x020402;
  SOUZU_3 = 0x020403;
  SOUZU_4 = 0x020404;
  SOUZU_5 = 0x020405;
  SOUZU_6 = 0x020406;
  SOUZU_7 = 0x020407;
  SOUZU_8 = 0x020408;
  SOUZU_9 = 0x020409;

  PINZU_1 = 0x020501;
  PINZU_2 = 0x020502;
  PINZU_3 = 0x020503;
  PINZU_4 = 0x020504;
  PINZU_5 = 0x020505;
  PINZU_6 = 0x020506;
  PINZU_7 = 0x020507;
  PINZU_8 = 0x020508;
  PINZU_9 = 0x020509;
}

enum HandElementType {
  MENTSU = 0x1;
    SHUNTSU = 0x11;
    KOUTSU = 0x12;
      KANTSU = 0x121;

  TOITSU = 0x2;
}

enum AgariType {
  RON = 0;
  TSUMO = 1;
}

enum RichiType {
  RICHI = 0;
  OPENED_RICHI = 1;
  DOUBLE_RICHI = 2;
}

enum MachiType {
  // 0x00 00 00 00 00 00 00 00
  //      G  F  E  D  C  B  A
  //
  // A: type
  // B: 0: any, 1: 2fu, 2: 0fu

  MASK_MACHI_KIND = 0xff;
  MASK_MACHI_FU = 0xff00;

  MACHI_2FU = 0x0100;
  MACHI_0FU = 0x0200;

  TANKI    = 0x0101;
  KANCHAN  = 0x0102;
  PENCHAN  = 0x0103;
  SHABO    = 0x0204;
  RYANMEN  = 0x0205;
  KYUMEN   = 0x0206;
  JUSANMEN = 0x0207;
}

enum PlayerType {
  DEALER = 0;
  NON_DEALER = 1;
}
